### ISSUE TITLE
[FEATURE REQUEST] Random user agent header

### ISSUE THREAD

**Issue by dsaxton** [epi052/feroxbuster#352]
**Is your feature request related to a problem? Please describe.**

Sometimes users may want to further obfuscate their scans by using a random user agent for each request. (This is a feature implemented for instance within `httpx` by passing `-random-agent`, and I think it might be handy to include in `feroxbuster` as well.)

**Describe the solution you'd like**

Some type of flag to randomize the user agent header.

**Describe alternatives you've considered**

There is the `--user-agent` option, but I'm not sure of a way to randomize this over realistic-looking values.

**Comment by epi052** [comment]
Howdy and thanks for the request!

This feature is doable. I had a look at what it would take to accomplish and it's a lot less effort than I initially thought it would be. 

- What user-agents would you want included? 
- Do they need to be random? Or, would a simple cycle over the user-agents, along with the inherent out-of-order requests due to async be enough randomness? 

In the interim, you can probably get what you want by passing feroxbuster through burp and playing with session handling rules.

**Comment by dsaxton** [comment]
Thanks for the quick response (and fantastic project)! I think some of the most common user agents might be a good approach: https://www.networkinghowtos.com/howto/common-user-agent-list/ (this link may be a little outdated but you get the idea). I'd also say random isn't necessary, and what you suggest seems like a good idea.

I'm still trying to think what would be the best way for users to specify this option. It could be its own binary flag (something like `--random-agent` or `--spoofed-agent`) which would essentially cause `--user-agent` to be ignored when set to true, or some type of special case wildcard value for `--user-agent`. Any preferences there (also if you wanted to point me to any relevant code blocks I could take a look at helping some myself)?

**Comment by epi052** [comment]
For the CLI option naming, if there's an existing convention among similar tools, I typically lean toward adopting the same where possible.  If `--random-agent` is what's used elsewhere, that's fine with me. 

For CLI parsing, I'd simply make `--random-agent` and `--user-agent` mutually exclusive. 

One could argue that a special wildcard aligns with the intent of #242. But I think for this particular case, I'm more of a fan of changing the behavior based on the options used instead of the values passed to the options.

If you're looking to work through a PR for this, that would be super! I'd be more than happy to work with you to get it in and working. 

If you wanted to take a look at adding `--random-agent` as a flag, below is my personal checklist for adding a new feature. Once those things are all done, then the feature itself would be the next step. If you have any questions about the checklist items, lmk, as it's for me, and may need a bit more info to be useful to someone else. 

# Add new command line option checklist
    - config/container.rs
        - [ ] add value to `Configuration` struct
        - [ ] add value to `Configuration::default`
        - [ ] update `Configuration` docstring
        - [ ] add value to `Configuration::parse_cli_args`
        - [ ] add value to `Configuration::merge_config`
    - config/tests.rs
        - [ ] update `Configuration::tests::setup_config_test` with new value
        - [ ] update `Configuration::tests::default_configuration` with new value
        - [ ] add test to check config parsing
    - ferox-config.toml.example
        - [ ] add new value
    - banner/container.rs
        - [ ] add new `BannerEntry` to `Banner` struct
        - [ ] update `Banner::new` with a `BannerEntry`
        - [ ] update `print_to` with whether to print or not
    - tests/test_banner.rs
        - [ ] add new test so that the new `BannerEntry` is displayed and examined under test
    - Cargo.toml
        - [ ] bump version - minor release +1 - patch release = 0
    - README.md
        - [ ] update default config section (if needed)
        - [ ] update example config file section
        - [ ] update help output section
        - [ ] consider adding to an example (new or existing)
        - [ ] update comparison table
    - parser.rs
        - [ ] add the option

**Comment by cortantief** [comment]
I will be interested too by that function.
I made a simple implementation using the convention of Mozilla and only using Firefox as web browser and windows, macos as operating system but we could easily add more, but for most use case I think it's enough.

That example use rand but one could think of another way to randomly create that User-Agent string if you don't want to have more crates.
Links: 
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4734c208edf7da625588fa16b5e2ed93
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent

**Comment by epi052** [comment]
As far as randomly choosing user-agents, i was thinking something like this:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1682b8ffa406f02f7cb32b1c2d2ddb71

We'd need to add a counter (not even a thread-safe counter necessarily). The inherent out-of-order-ness of async networking code should add randomness. We'll know the UA's ahead of time and can just have a static list and select based on modding the counter. 

If it doesn't work out, we can add rand, but i don't think it'll be necessary

**Comment by epi052** [comment]
For the implementation of actually overriding the default header, i was able to use  https://docs.rs/reqwest/0.11.4/reqwest/struct.RequestBuilder.html#method.header to override the existing user-agent that gets baked into the client. 

`client.get` returns a `RequestBuilder`. You should be able to modify https://github.com/epi052/feroxbuster/blob/main/src/utils.rs#L133 to get the modified user-agent into each request. 

I THINK that's the only location from which we make meaningful requests to the server (heuristics makes 1-3 or 4 depending).

**Comment by dsaxton** [comment]
I agree we'd probably not want to add a dependency on `rand` for this only. @epi052 I actually started working on your checklist, should I push that up now (I will very likely get stuck and need help, or you could just push straight to the branch if you already have the implementation figured out)?

**Comment by epi052** [comment]
go ahead and submit the PR, we can mind-meld from there.  I haven't thought through everything, but I checked to make sure we could override the user-agent after a `reqwest::Client` was created.
